Add documentation for CA1870 (#37056)
dotnet/docs: docs/fundamentals/code-analysis/quality-rules/ca1870.md

---
title: "CA1870: Use a cached 'SearchValues' instance"
description: "Learn about code analyzer rule CA1870 - Use a cached 'SearchValues' instance"
ms.date: 09/12/2023
ms.topic: reference
f1_keywords:
 - CA1870
 - UseSearchValuesAnalyzer
helpviewer_keywords:
 - CA1870
dev_langs:
 - CSharp
author: MihaZupan
ms.author: mizupan
---

# CA1870: Use a cached 'SearchValues' instance

| Property                            | Value                                  |
|-------------------------------------|----------------------------------------|
| **Rule ID**                         | CA1870                                 |
| **Title**                           | Use a cached 'SearchValues' instance   |
| **Category**                        | [Performance](performance-warnings.md) |
| **Fix is breaking or non-breaking** | Non-breaking                           |

> | **Enabled by default in .NET 7**    | No                                     |

Author: I copied this from existing CA docs, does it need changing?

This analyzer is being added in .NET 8, where it should be enabled by default.

Reviewer: Yes, let's update it to .NET 8.

## Cause

> An `IndexOfAny`/`ContainsAny` method is being called with many constant values in a way that can benefit from using <xref:System.Buffers.SearchValues> instead.

Reviewer: ```suggestion
An `IndexOfAny` or `ContainsAny` method is called with many constant values in a way that can benefit from using <xref:System.Buffers.SearchValues> instead.
```

> The rule will not flag calls using up to 5 values, as those are already using an optimal implementation.

Reviewer: ```suggestion
The rule doesn't flag calls that use up to five values, as those already use an optimal implementation.
```

## Rule description

> Using a cached <xref:System.Buffers.SearchValues%601> instance is more efficient than passing values to `IndexOfAny`/`ContainsAny` directly.

Reviewer: ```suggestion
Using a cached <xref:System.Buffers.SearchValues%601> instance is more efficient than passing values to `IndexOfAny` or `ContainsAny` directly.
```

## How to fix violations

> Create and cache a <xref:System.Buffers.SearchValues%601> instance in a `static readonly` field, then pass that instance to the `IndexOfAny`/`ContainsAny` call instead.

Reviewer: ```suggestion
Create and cache a <xref:System.Buffers.SearchValues%601> instance in a `static readonly` field, then pass that instance to the `IndexOfAny` or `ContainsAny` call instead.
```
